File: internal/timeline/timeline.go
// Package timeline is the domain model the frontend consumes: a virtual-clock
// ordered list of normalized scheduler/GC events plus run metadata. It is the
// JSON contract between the backend and the (future) PixiJS player, and it knows
// nothing about how traces are produced or parsed.
package timeline

// EventType enumerates the normalized animation events.
type EventType string

const (
	EventGCreate       EventType = "g_create"
	EventGRunStart     EventType = "g_run_start"
	EventGRunStop      EventType = "g_run_stop"
	EventGBlock        EventType = "g_block"
	EventGUnblock      EventType = "g_unblock"
	EventGSyscallEnter EventType = "g_syscall_enter"
	EventGSyscallExit  EventType = "g_syscall_exit"
	EventGExit         EventType = "g_exit"
	EventPStart        EventType = "p_start"
	EventPStop         EventType = "p_stop"
	EventGCRangeBegin  EventType = "gc_range_begin"
	EventGCRangeEnd    EventType = "gc_range_end"
	EventMetric        EventType = "metric"
)

// NoResource marks Event.GID / Event.PID / Event.MID when the field does not
// apply. We use -1 (not omitempty) because goroutine 0, proc 0 and thread M0
// are all valid IDs and would be wrongly dropped by omitempty.
const NoResource int64 = -1

// Event is one normalized point on the timeline.
//
// MID is the OS thread (M) of the *executing context* that emitted the event,
// verbatim from the trace. Careful: on g_unblock/g_create it is the
// unblocker's/creator's M (not the target goroutine's), and on a p_stop caused
// by a steal it is the stealer's M. Consumers must bind an M to a G/P only on
// own-execution events (g_run_start, g_syscall_enter/exit, p_start).
type Event struct {
	T      int64     `json:"t"`    // ns since the first trace event
	Type   EventType `json:"type"` //
	GID    int64     `json:"gid"`  // goroutine id, or NoResource
	PID    int64     `json:"pid"`  // proc id, or NoResource
	MID    int64     `json:"mid"`  // executing OS-thread id, or NoResource
	Reason string    `json:"reason,omitempty"`
	Name   string    `json:"name,omitempty"`  // range / metric name
	Value  uint64    `json:"value,omitempty"` // metric value
	Stolen bool      `json:"stolen,omitempty"`
}

// Meta describes the run as a whole.
type Meta struct {
	Scenario   string  `json:"scenario"`
	NumProcs   int     `json:"numProcs"`
	DurationNs int64   `json:"durationNs"`
	Goroutines []int64 `json:"goroutines"`
}

// Timeline is the full payload sent to the frontend.
type Timeline struct {
	Meta   Meta    `json:"meta"`
	Events []Event `json:"events"`
}

File: web/src/model/timeline.ts
// Mirror of the Go timeline DTO (internal/timeline). Keep these in sync with the
// backend's JSON contract.

export type EventType =
  | 'g_create'
  | 'g_run_start'
  | 'g_run_stop'
  | 'g_block'
  | 'g_unblock'
  | 'g_syscall_enter'
  | 'g_syscall_exit'
  | 'g_exit'
  | 'p_start'
  | 'p_stop'
  | 'gc_range_begin'
  | 'gc_range_end'
  | 'metric'

/** Sentinel for an absent goroutine/proc/thread id (matches timeline.NoResource). */
export const NO_RESOURCE = -1

export interface TimelineEvent {
  /** ns since the first trace event. */
  t: number
  type: EventType
  gid: number
  pid: number
  /**
   * OS thread (M) of the *executing context*, verbatim from the trace.
   * Careful: on g_unblock/g_create it is the unblocker's/creator's M (not the
   * target goroutine's), and on a steal-caused p_stop it is the stealer's M.
   * Bind an M to a G/P only on own-execution events — see stateAt.
   */
  mid: number
  reason?: string
  name?: string
  value?: number
  stolen?: boolean
}

export interface Meta {
  scenario: string
  numProcs: number
  durationNs: number
  goroutines: number[]
}

export interface Timeline {
  meta: Meta
  events: TimelineEvent[]
}

export interface ScenarioParam {
  name: string
  min: number
  max: number
  default: number
}

export interface ScenarioInfo {
  id: string
  title: string
  description: string
  order: number
  params: ScenarioParam[]
}
